articles/changes.md

```diff
@@ -0,0 +1,257 @@
+In ROS 2 a Python package can use anything in setup.py files, e.g. entry points since they are being invoked with `python3 setup.py install`.
+
+
+### Environment setup
+
+In ROS 1 the build tool generates scripts which must be sourced in order to setup the environment before being able to use the built ROS packages.
+This approach only works when the ROS packages are being built with ROS specific build tool.
+
+In ROS 2 the environment setup is separated into package-specific scripts and workspace-specific scripts.
+Each package provides the necessary scripts to make itself usable after being built.
+The build tool only invokes the workspace-specific scripts which then call the package-specific scripts.
+
+
+### No non-isolated build
+
+In ROS 1 multiple packages can be built in a single CMake context.
+While this speeds up the build step, every package needs to ensure that cross package target dependencies are defined correctly.
+Additionally all packages share the same namespace which leads to colliding target names, etc.
+
+In ROS 2 only isolated builds are supported, i.e. every package is built separately on its own.
+
+
+### No devel space
+
+In ROS 1 packages can be built without installing them.
+From the devel space in combination with the source space the system is already usable.
+But every package has to actively support the devel space, e.g. in environment hooks and CMake code.
+
+In ROS 2 a package must be installed after building it before it can be used.
+
+One reason for the devel space in ROS 1 is to enable the developer to change files, e.g. Python code or launch files, and use the modified code directly without the need to rebuild the package.
+This benefit is preserved in ROS 2 by optionally replacing copy operations in the install steps with symlinks.
+
+
+### Support catkin_simple use case
+
+In ROS 1 the package catkin_simple is aiming to make writing the CMake code of ROS packages easier.
+In many cases it is not achieving this goal which is often due to restrictions of the design necessary for support features like the devel space.
+
+In ROS 2 the CMake API was restructured to support this use case.
+
+
+### Minimal support for packages without manifest
+
+In ROS 1 only packages with a manifest file are considered by the build system.
+In ROS 2 it is possible to detect packages with supported build system in folders without a manifest file.
+If the package follows common practice it might even be possible to detect some of the missing meta information (like dependencies).
+
+
+## Messages, Services
+
+For more information please see the [ROS interface definition](http://design.ros2.org/articles/interface_definition.html) article.
+
+
+### Separated namespaces
+
+In ROS 1 .msg and .srv files can have the same name but the generated code collides.
+The same is the case for the request and response parts of services.
+
+In ROS 2 the generated code uses separate namespaces to guarantee it is collision-free.
```

Comment:
+1 Good idea, @wjwwood 

I didn't know about the ``std_svcs/Empty.h`` issue, but it supports my point that there is *no* *way* to have a name collision between a ROS 1 message and a service. As a purely architectural issue, I don't think we should introduce that as a possibility.

Nor do I see a need for ``std_msgs/Empty.h`` to throw a compile error. That name unambiguously refers to the message. If the service had been desired, the program would have included ``std_svcs/Empty.h`` instead.

> We can choose whether or not to put a deprecation message in the compat header to encourage people to migrate to the new format, which I still believe is better because it is more conceptually consistent and provides more information to people reading unfamiliar code.

Please don't deprecate those references any time in the first few years. Maybe later, after most code has finally been migrated.

> We can provide similar symbols to what is in ROS 1, but somethings will likely end up being different, so I'm still not optimistic that we can achieve a zero change migration w.r.t. messages.

I don't expect zero change. The reason I keep harping on these message incompatibilities is because most ROS 1 code I know about uses messages all over the place as fundamental data structures. It's very convenient to do it that way.

I want to eliminate the bulk of these message incompatibilities, so we'll be looking at 50-line diffs for a package migration, and not 500-line diffs. That order of magnitude will save engineer-centuries of work for the entire ROS community.